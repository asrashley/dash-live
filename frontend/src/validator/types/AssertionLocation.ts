export interface AssertionLocation {
    filename: string;
    line: number;
}
