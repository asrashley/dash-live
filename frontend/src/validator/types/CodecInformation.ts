export interface CodecDetails {
    label: string;
    error?: string;
    details: string[];
}

export interface CodecInformation {
    codec: string;
    error?: string;
    details: CodecDetails[];
}
