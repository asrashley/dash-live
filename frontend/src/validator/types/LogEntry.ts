export type LogEntry = {
    id: number;
    level: 'error' | 'info' | 'debug';
    text: string;
};
