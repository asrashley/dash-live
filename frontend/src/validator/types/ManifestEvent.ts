
export interface ManifestEvent {
    text: string[];
}