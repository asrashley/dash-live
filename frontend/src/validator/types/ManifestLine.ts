export interface ManifestLine {
    line: number;
    text: string;
    hasError: boolean;
    errors: string[];
}
