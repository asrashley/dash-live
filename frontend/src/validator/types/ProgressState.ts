export interface ProgressState {
    currentValue?: number;
    minValue: number;
    maxValue: number;
    text: string;
    finished?: boolean;
    error?: boolean;
 }
