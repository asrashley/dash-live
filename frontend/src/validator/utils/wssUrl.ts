declare const _SERVER_PORT_: number | null;

export interface WssUrlProps {
  protocol: URL["protocol"];
  hostname: URL["hostname"];
  port: URL["port"];
}
export function wssUrl({ protocol, hostname, port }: WssUrlProps): string {
  const wsProto = protocol === 'http:' ? 'ws://' : 'wss://';
  return `${wsProto}${hostname}:${_SERVER_PORT_ ?? port}/`;
}
