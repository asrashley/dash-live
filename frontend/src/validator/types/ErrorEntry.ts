import { AssertionLocation } from "./AssertionLocation";

export interface ErrorEntry {
    assertion: AssertionLocation;
    location: [number, number];
    clause?: string;
    msg: string;
}