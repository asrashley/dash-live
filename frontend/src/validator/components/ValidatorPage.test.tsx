import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { io, Socket } from "socket.io-client";
import { mock, mockReset } from "vitest-mock-extended";
import userEvent from "@testing-library/user-event";
import log from "loglevel";

import { wssUrl } from "../utils/wssUrl";
import { FakeEndpoint } from "../../test/FakeEndpoint";
import { MockWebsocketServer } from "../../test/MockWebsocketServer";
import ProtectedValidatorPage, { ValidatorPage } from "./ValidatorPage";
import { renderWithFormAccess } from "../test/renderWithFormAccess";
import { ApiRequests, EndpointContext } from "../../endpoints";
import { LoginResponse } from "../../types/LoginResponse";
import { mediaUser } from "../../test/MockServer";
import { renderWithProviders } from "../../test/renderWithProviders";
import { act } from "@testing-library/preact";

vi.mock("socket.io-client");

vi.mock("../utils/wssUrl");

describe("ValidatorPage component", () => {
  const manifest = "http://localhost:8765/dash/vod/bbb/hand_made.mpd";
  const websocketUrl = "wss://localhost:3456";
  const wssUrlMock = vi.mocked(wssUrl);
  const ioMock = vi.mocked(io);
  const mockSocket = mock<Socket>();
  let endpoint: FakeEndpoint;
  let server: MockWebsocketServer;

  beforeEach(() => {
    wssUrlMock.mockReturnValue(websocketUrl);
    ioMock.mockImplementation(() => mockSocket);
    endpoint = new FakeEndpoint(websocketUrl);
    server = new MockWebsocketServer(mockSocket, endpoint);
    mockSocket.connect.mockImplementation(server.connect);
    mockSocket.disconnect.mockImplementation(server.disconnect);
    mockSocket.on.mockImplementation(server.on);
    mockSocket.off.mockImplementation(server.off);
    mockSocket.emit.mockImplementation(server.emit);
  });

  afterEach(async () => {
    await server.destroy();
    vi.clearAllMocks();
    mockReset(mockSocket);
    log.setLevel("error");
  });

  test("initial state", async () => {
    const { asFragment, getByTestId, startBtn, cancelBtn } =
      renderWithFormAccess(<ValidatorPage />);
    await expect(server.getConnectedPromise()).resolves.toBeUndefined();
    expect(startBtn.disabled).toEqual(true);
    expect(cancelBtn.disabled).toEqual(true);
    const stateElt = getByTestId("validator-state-badge") as HTMLElement;
    expect(stateElt.innerHTML).toEqual("idle");
    expect(ioMock).toHaveBeenCalledTimes(1);
    expect(ioMock).toHaveBeenCalledWith(websocketUrl, {
      autoConnect: false,
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test("can validate a stream", async () => {
    const userEv = userEvent.setup();

    const { getByTestId, manifestElt, startBtn } = renderWithFormAccess(
      <ValidatorPage />
    );
    await userEv.click(manifestElt);
    await userEv.clear(manifestElt);
    await userEv.type(manifestElt, manifest);
    expect(startBtn.disabled).toEqual(false);
    const done = server.getDonePromise();
    await userEv.click(startBtn);
    let finished = false;
    while (!finished) {
      await act(async () => {
        finished = await server.nextTick(0.5);
      });
    }
    await expect(done).resolves.toBeUndefined();
    const stateElt = getByTestId("validator-state-badge") as HTMLElement;
    expect(stateElt.innerHTML).toEqual("done");
  });

  describe("ProtectedValidatorPage component", () => {
    const apiRequests = mock<ApiRequests>();

    afterEach(() => {
      vi.clearAllMocks();
      mockReset(apiRequests);
    });

    test("initial state, not logged in", async () => {
      apiRequests.getUserInfo.mockImplementation(
        async () => new Response(null, { status: 401 })
      );
      const { asFragment, findByLabelText } = renderWithProviders(
        <EndpointContext.Provider value={apiRequests}>
          <ProtectedValidatorPage />
        </EndpointContext.Provider>
      );
      await expect(server.getConnectedPromise()).resolves.toBeUndefined();
      const saveElt = (await findByLabelText(
        "Add stream to this server?"
      )) as HTMLInputElement;
      expect(saveElt.disabled).toEqual(true);
      expect(asFragment()).toMatchSnapshot();
    });

    test("initial state, logged in user", async () => {
      const login: LoginResponse = {
        success: true,
        csrf_token: "123",
        user: mediaUser,
      };
      apiRequests.getUserInfo.mockImplementation(async () => login);
      const { asFragment, findByLabelText } = renderWithProviders(
        <EndpointContext.Provider value={apiRequests}>
          <ProtectedValidatorPage />
        </EndpointContext.Provider>,
        { userInfo: mediaUser }
      );
      await expect(server.getConnectedPromise()).resolves.toBeUndefined();
      const saveElt = (await findByLabelText(
        "Add stream to this server?"
      )) as HTMLInputElement;
      expect(saveElt.disabled).toEqual(false);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
