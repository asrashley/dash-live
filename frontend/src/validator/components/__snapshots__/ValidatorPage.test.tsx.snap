// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ValidatorPage component > ProtectedValidatorPage component > initial state, logged in user 1`] = `
<DocumentFragment>
  <div
    class="container"
    id="validator"
  >
    <form
      class="card"
      id="manifest-form"
      name="manifest"
    >
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-manifest"
          id="label-manifest"
        >
          Manifest to check:
        </label>
        <div
          class="col-10"
        >
          <input
            aria-describedby="label-manifest"
            class="form-control is-invalid "
            id="model-manifest"
            name="manifest"
            placeholder="... manifest URL ..."
            title="Manifest to check"
            type="url"
          />
        </div>
        <div
          class="invalid-feedback d-block"
        >
          manifest URL is required
        </div>
      </div>
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-duration"
          id="label-duration"
        >
          Maximum duration:
        </label>
        <div
          class="col-7"
        >
          <input
            aria-describedby="text-duration"
            class="form-control is-valid "
            id="model-duration"
            name="duration"
            title="Maximum duration"
            type="number"
          />
        </div>
        <div
          class="col-3 form-text"
          id="text-duration"
        >
          seconds
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-prefix"
          id="label-prefix"
        >
          Destination directory:
        </label>
        <div
          class="col-7"
        >
          <input
            aria-describedby="text-prefix"
            class="form-control is-valid "
            disabled=""
            id="model-prefix"
            name="prefix"
            title="Destination directory"
            type="text"
          />
        </div>
        <div
          class="col-3 form-text"
          id="text-prefix"
        >
          3 to 31 characters without any special characters
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-title"
          id="label-title"
        >
          Stream title:
        </label>
        <div
          class="col-10"
        >
          <input
            aria-describedby="label-title"
            class="form-control is-valid "
            disabled=""
            id="model-title"
            name="title"
            title="Stream title"
            type="text"
          />
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <div
          class="col-12"
        >
          <div
            class="form-check form-switch is-valid "
            data-testid="msi-row1"
          >
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msiencrypted"
                name="encrypted"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msiencrypted"
              >
                Stream is encrypted?
              </label>
            </div>
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msimedia"
                name="media"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msimedia"
              >
                Check media segments
              </label>
            </div>
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msiverbose"
                name="verbose"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msiverbose"
              >
                Verbose output
              </label>
            </div>
          </div>
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <div
          class="col-12"
        >
          <div
            class="form-check form-switch is-valid "
            data-testid="msi-row2"
          >
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msipretty"
                name="pretty"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msipretty"
              >
                Pretty print XML before validation
              </label>
            </div>
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msisave"
                name="save"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msisave"
              >
                Add stream to this server?
              </label>
            </div>
          </div>
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="form-actions mt-2"
      >
        <button
          class="btn btn-primary"
          disabled=""
        >
          Validate DASH stream
        </button>
        <button
          class="btn btn-secondary"
          disabled=""
          id="btn-cancel"
        >
          Cancel
        </button>
      </div>
    </form>
    <div
      class="position-relative"
    >
      <div
        class="card progress"
      >
        <div>
          <div
            aria-valuemax="100"
            aria-valuemin="0"
            class="progress-bar progress-bar-striped visually-hidden"
            role="progressbar"
          >
            NaN%
          </div>
          <div
            class="progress-text"
          />
        </div>
      </div>
      <div
        class="position-absolute top-0 start-50 translate-middle badge rounded-pill text-bg-secondary"
        data-testid="validator-state-badge"
      >
        idle
      </div>
    </div>
    <div
      class="card"
      id="manifest-text"
    />
    <table
      class="d-none"
    >
      <thead>
        <tr>
          <th
            class="lineno"
          >
            MPD line
          </th>
          <th
            class="clause"
          >
            DASH clause
          </th>
          <th
            class="message"
          >
            Error Details
          </th>
          <th
            class="source-location"
          >
            Source code
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <table
      class="d-none"
    >
      <thead>
        <tr>
          <th
            class="codec-string"
          >
            Codec
          </th>
          <th
            class="codec-detail"
          >
            Details
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <div
      class="card results"
    />
  </div>
</DocumentFragment>
`;

exports[`ValidatorPage component > ProtectedValidatorPage component > initial state, not logged in 1`] = `
<DocumentFragment>
  <div
    class="container"
    id="validator"
  >
    <form
      class="card"
      id="manifest-form"
      name="manifest"
    >
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-manifest"
          id="label-manifest"
        >
          Manifest to check:
        </label>
        <div
          class="col-10"
        >
          <input
            aria-describedby="label-manifest"
            class="form-control is-invalid "
            id="model-manifest"
            name="manifest"
            placeholder="... manifest URL ..."
            title="Manifest to check"
            type="url"
          />
        </div>
        <div
          class="invalid-feedback d-block"
        >
          manifest URL is required
        </div>
      </div>
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-duration"
          id="label-duration"
        >
          Maximum duration:
        </label>
        <div
          class="col-7"
        >
          <input
            aria-describedby="text-duration"
            class="form-control is-valid "
            id="model-duration"
            name="duration"
            title="Maximum duration"
            type="number"
          />
        </div>
        <div
          class="col-3 form-text"
          id="text-duration"
        >
          seconds
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-prefix"
          id="label-prefix"
        >
          Destination directory:
        </label>
        <div
          class="col-7"
        >
          <input
            aria-describedby="text-prefix"
            class="form-control is-valid "
            disabled=""
            id="model-prefix"
            name="prefix"
            title="Destination directory"
            type="text"
          />
        </div>
        <div
          class="col-3 form-text"
          id="text-prefix"
        >
          3 to 31 characters without any special characters
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-title"
          id="label-title"
        >
          Stream title:
        </label>
        <div
          class="col-10"
        >
          <input
            aria-describedby="label-title"
            class="form-control is-valid "
            disabled=""
            id="model-title"
            name="title"
            title="Stream title"
            type="text"
          />
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <div
          class="col-12"
        >
          <div
            class="form-check form-switch is-valid "
            data-testid="msi-row1"
          >
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msiencrypted"
                name="encrypted"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msiencrypted"
              >
                Stream is encrypted?
              </label>
            </div>
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msimedia"
                name="media"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msimedia"
              >
                Check media segments
              </label>
            </div>
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msiverbose"
                name="verbose"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msiverbose"
              >
                Verbose output
              </label>
            </div>
          </div>
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <div
          class="col-12"
        >
          <div
            class="form-check form-switch is-valid "
            data-testid="msi-row2"
          >
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msipretty"
                name="pretty"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msipretty"
              >
                Pretty print XML before validation
              </label>
            </div>
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                disabled=""
                id="msisave"
                name="save"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msisave"
              >
                Add stream to this server?
              </label>
            </div>
          </div>
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="form-actions mt-2"
      >
        <button
          class="btn btn-primary"
          disabled=""
        >
          Validate DASH stream
        </button>
        <button
          class="btn btn-secondary"
          disabled=""
          id="btn-cancel"
        >
          Cancel
        </button>
      </div>
    </form>
    <div
      class="position-relative"
    >
      <div
        class="card progress"
      >
        <div>
          <div
            aria-valuemax="100"
            aria-valuemin="0"
            class="progress-bar progress-bar-striped visually-hidden"
            role="progressbar"
          >
            NaN%
          </div>
          <div
            class="progress-text"
          />
        </div>
      </div>
      <div
        class="position-absolute top-0 start-50 translate-middle badge rounded-pill text-bg-secondary"
        data-testid="validator-state-badge"
      >
        idle
      </div>
    </div>
    <div
      class="card"
      id="manifest-text"
    />
    <table
      class="d-none"
    >
      <thead>
        <tr>
          <th
            class="lineno"
          >
            MPD line
          </th>
          <th
            class="clause"
          >
            DASH clause
          </th>
          <th
            class="message"
          >
            Error Details
          </th>
          <th
            class="source-location"
          >
            Source code
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <table
      class="d-none"
    >
      <thead>
        <tr>
          <th
            class="codec-string"
          >
            Codec
          </th>
          <th
            class="codec-detail"
          >
            Details
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <div
      class="card results"
    />
  </div>
</DocumentFragment>
`;

exports[`ValidatorPage component > initial state 1`] = `
<DocumentFragment>
  <div
    class="container"
    id="validator"
  >
    <form
      class="card"
      id="manifest-form"
      name="manifest"
    >
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-manifest"
          id="label-manifest"
        >
          Manifest to check:
        </label>
        <div
          class="col-10"
        >
          <input
            aria-describedby="label-manifest"
            class="form-control is-invalid "
            id="model-manifest"
            name="manifest"
            placeholder="... manifest URL ..."
            title="Manifest to check"
            type="url"
          />
        </div>
        <div
          class="invalid-feedback d-block"
        >
          manifest URL is required
        </div>
      </div>
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-duration"
          id="label-duration"
        >
          Maximum duration:
        </label>
        <div
          class="col-7"
        >
          <input
            aria-describedby="text-duration"
            class="form-control is-valid "
            id="model-duration"
            name="duration"
            title="Maximum duration"
            type="number"
          />
        </div>
        <div
          class="col-3 form-text"
          id="text-duration"
        >
          seconds
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-prefix"
          id="label-prefix"
        >
          Destination directory:
        </label>
        <div
          class="col-7"
        >
          <input
            aria-describedby="text-prefix"
            class="form-control is-valid "
            disabled=""
            id="model-prefix"
            name="prefix"
            title="Destination directory"
            type="text"
          />
        </div>
        <div
          class="col-3 form-text"
          id="text-prefix"
        >
          3 to 31 characters without any special characters
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <label
          class="col-2 col-form-label"
          for="model-title"
          id="label-title"
        >
          Stream title:
        </label>
        <div
          class="col-10"
        >
          <input
            aria-describedby="label-title"
            class="form-control is-valid "
            disabled=""
            id="model-title"
            name="title"
            title="Stream title"
            type="text"
          />
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <div
          class="col-12"
        >
          <div
            class="form-check form-switch is-valid "
            data-testid="msi-row1"
          >
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msiencrypted"
                name="encrypted"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msiencrypted"
              >
                Stream is encrypted?
              </label>
            </div>
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msimedia"
                name="media"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msimedia"
              >
                Check media segments
              </label>
            </div>
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msiverbose"
                name="verbose"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msiverbose"
              >
                Verbose output
              </label>
            </div>
          </div>
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="row mb-2 form-group "
      >
        <div
          class="col-12"
        >
          <div
            class="form-check form-switch is-valid "
            data-testid="msi-row2"
          >
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                id="msipretty"
                name="pretty"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msipretty"
              >
                Pretty print XML before validation
              </label>
            </div>
            <div
              class="form-check form-check-inline"
            >
              <input
                class="form-check-input"
                disabled=""
                id="msisave"
                name="save"
                type="checkbox"
              />
              <label
                class="form-check-label me-3"
                for="msisave"
              >
                Add stream to this server?
              </label>
            </div>
          </div>
        </div>
        <div
          class="invalid-feedback d-none"
        />
      </div>
      <div
        class="form-actions mt-2"
      >
        <button
          class="btn btn-primary"
          disabled=""
        >
          Validate DASH stream
        </button>
        <button
          class="btn btn-secondary"
          disabled=""
          id="btn-cancel"
        >
          Cancel
        </button>
      </div>
    </form>
    <div
      class="position-relative"
    >
      <div
        class="card progress"
      >
        <div>
          <div
            aria-valuemax="100"
            aria-valuemin="0"
            class="progress-bar progress-bar-striped visually-hidden"
            role="progressbar"
          >
            NaN%
          </div>
          <div
            class="progress-text"
          />
        </div>
      </div>
      <div
        class="position-absolute top-0 start-50 translate-middle badge rounded-pill text-bg-secondary"
        data-testid="validator-state-badge"
      >
        idle
      </div>
    </div>
    <div
      class="card"
      id="manifest-text"
    />
    <table
      class="d-none"
    >
      <thead>
        <tr>
          <th
            class="lineno"
          >
            MPD line
          </th>
          <th
            class="clause"
          >
            DASH clause
          </th>
          <th
            class="message"
          >
            Error Details
          </th>
          <th
            class="source-location"
          >
            Source code
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <table
      class="d-none"
    >
      <thead>
        <tr>
          <th
            class="codec-string"
          >
            Codec
          </th>
          <th
            class="codec-detail"
          >
            Details
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <div
      class="card results"
    />
  </div>
</DocumentFragment>
`;
